src/osdc/Journaler.cc

```diff
@@ -493,7 +493,21 @@ void Journaler::_finish_write_head(int r, Header &wrote,
   }
   ceph_assert(!readonly);
   ldout(cct, 10) << "_finish_write_head " << wrote << dendl;
-  last_committed = wrote;
+  if (wrote.write_pos < last_committed.write_pos ||
+      wrote.expire_pos < last_committed.expire_pos ||
+      wrote.trimmed_pos < last_committed.trimmed_pos) {
+    lderr(cct) << __func__ << ": not updating last_committed: "
+	       << "(wrote.write_pos/last_committed.write_pos="
+	       << wrote.write_pos << "," << last_committed.write_pos < "), "
+	       << "(wrote.expire_pos/last_committed.expire_pos="
+	       << wrote.expire_pos << "," << last_committed.expire_pos < "), "
+	       << "(wrote.trimmed_pos/last_committed.trimmed_pos="
+	       << wrote.trimmed_pos << "," << last_committed.trimmed_pos < ")"
+	       << dendl;
+    ceph_abort();
+  } else {
+    last_committed = wrote;
+  }
```

Comment:
With this change, the old values are already written to the disk and we have no clue whether the write op is out of sync or the just the update notification or both. But the log should help as a reference for setting the header field values, using `cephfs-journal-tool`, when the assertion failure `ceph_assert(trim_to >= trimming_pos)` is hit.

Reply:
but the assertion failure in `MDLog::_trim()` will not happen due to the `ceph_abort()` here

Reply:
Yeh, right. This is just a hunch and right now we have no idea what could have happened that caused the journal pointer values.